### ISSUE TITLE
"cannot be assigned to the mutable type 'Type<unknown>[]'" error appears when createUnionType

### ISSUE THREAD

**Issue by yiuhei-AMPD** [nestjs/graphql#2122]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

When I follow the code first example in the following document to create a union type, 
it returns the following error
```
error TS4104: The type 'readonly [typeof Author, typeof Book]' is 'readonly' and cannot be assigned to the mutable type 'Type<unknown>[]'.

105   types: () => [Author, Book] as const,
                   ~~~~~~~~~~~~~~~~~~~~~~~

  node_modules/@nestjs/graphql/dist/type-factories/create-union-type.factory.d.ts:7:12
    7     types: () => T;
                 ~~~~~~~
    The expected type comes from the return type of this signature.
    
```

Document:
https://docs.nestjs.com/graphql/unions-and-enums

Code:
```
export const ResultUnion = createUnionType({
  name: 'ResultUnion',
  types: () => [Author, Book] as const,
});
 ```

### Minimum reproduction code

https:github.com

### Steps to reproduce

Add the follow code which copy from the document to the graphql model
```
@ObjectType()
export class Book {
  @Field()
  title: string;
}

@ObjectType()
export class Author {
  @Field()
  name: string;
}

export const ResultUnion = createUnionType({
  name: 'ResultUnion',
  types: () => [Author, Book] as const,
});
```

### Expected behavior

No error appears

### Package version

8.1.3

### Graphql version

`graphql`: 
`apollo-server-express`: 
`apollo-server-fastify`: 

### In which operating systems have you tested?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by kamilmysliwiec** [comment]
I can't reproduce your issue.

Please provide a [minimum reproduction repository](https://minimum-reproduction.wtf/) (Git repository/[StackBlitz](https://stackblitz.com/fork/github/nestjs/typescript-starter)/[CodeSandbox](https://codesandbox.io/s/github/nestjs/typescript-starter/tree/master) project).

**Comment by ashleyww93** [comment]
I am also getting this issue.

<img width="1056" alt="image" src="https://user-images.githubusercontent.com/3884418/162723765-d44720cd-d8bf-4fcf-a331-5c108a80181d.png">

**Comment by ashleyww93** [comment]
I got this working, by doing the following:

```
import { createUnionType } from '@nestjs/graphql';
import { CloudshelfThemeClassicAttractScreen } from './cloudshelf-theme-classic-attract-screen';
import { CloudshelfThemeMinimalAttractScreen } from './cloudshelf-theme-minimal-attract-screen';

export const AttractScreenOptionsUnion = createUnionType({
    name: 'AttractScreenOptionsUnion',
    types: () => [CloudshelfThemeClassicAttractScreen, CloudshelfThemeMinimalAttractScreen],
    resolveType(value) {
        if (typeof value.displayFrame !== undefined) {
            return CloudshelfThemeClassicAttractScreen;
        }
        if (typeof value.actionButton !== undefined) {
            return CloudshelfThemeMinimalAttractScreen;
        }
        return null;
    },
});

```

**Comment by kamilmysliwiec** [comment]
Please provide a [minimum reproduction repository](https://minimum-reproduction.wtf/) (Git repository/[StackBlitz](https://stackblitz.com/fork/github/nestjs/typescript-starter)/[CodeSandbox](https://codesandbox.io/s/github/nestjs/typescript-starter/tree/master) project).

**Comment by ashleyww93** [comment]
Read. The. Comment.

I don't need to provide a minimum reproduction repository; I had the same issue as the OP, did my own investigation because the nestjs docs are out of date(or just bad), and provided a fix for the OP. 

Multiple people are having the issue when following the docs, so clearly they are wrong.

**Comment by kamilmysliwiec** [comment]
@ashleyww93 having a minimum reproduction repository (or at least a minimal Stackblitz/Code sandbox project) would simply help us investigate & fix this issue quicker. We have e2e tests in this repository where we use `createUnionType` multiple times + the repository uses the latest version of TypeScript (4.6.3) and this error does not occur. This must be a combination of a specific TS version + `@nestjs/graphql` version + usage which is just hard to reproduce

**Comment by mdrideout** [comment]
I am also having this issue. Minimum reproduction repository here, see the union module / types.
https://stackblitz.com/edit/nestjs-typescript-starter-nuxmjd?file=src/union-example/types/type-union.ts

I am not getting the error in that environment though, the error occurs in VSCode where I have ensured the tsconfig.json is identical to the sample. I've also ensured that my local nestjs and the environments versions are the same.

Screenshot of error in VSCode environment
![Screen Shot 2022-04-14 at 11 21 43 AM](https://user-images.githubusercontent.com/3958274/163421809-9aea7fe6-d777-4be7-a366-c31e34a3b8ba.png)


In the VSCode environment, the error goes away if I remove `as const`, unsure if this will cause unintended side effects. But NestJS does appear to generate schema acceptable to Apollo Studio with that omission. I am also able to successfully run the query from the Apollo Studio client. In the example, type-union.ts object would become:
```ts
export const UnionType = createUnionType({
  name: 'UnionType',
  types: () => [FirstType, SecondType],
});
```

My project's full package.json for reference
```json
{
  "name": "graphql-server",
  "version": "0.0.1",
  "description": "",
  "author": "",
  "private": true,
  "license": "UNLICENSED",
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "nest build",
    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
    "start": "nest start",
    "start:local": "STAGE=local nest start --watch",
    "start:debug": "STAGE=local nest start --debug --watch",
    "start:staging": "STAGE=staging nest start --debug --watch",
    "start:prod": "STAGE=prod node dist/main",
    "start:azure": "SERVER=azure npm run build && func start --port 7072",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "test": "STAGE=local jest",
    "test:watch": "STAGE=local jest --watch --silent",
    "test:watch:verbose": "STAGE=local jest --watch",
    "test:cov": "jest --coverage",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test:e2e": "jest --config ./test/jest-e2e.json"
  },
  "dependencies": {
    "@azure/functions": "^1.0.3",
    "@azure/identity": "^1.5.1",
    "@azure/storage-blob": "^12.7.0",
    "@azure/cosmos": "^3.15.1",
    "@nestjs/axios": "^0.0.1",
    "@nestjs/azure-func-http": "^0.7.0",
    "@nestjs/common": "^8.0.0",
    "@nestjs/config": "^1.0.1",
    "@nestjs/core": "^8.0.0",
    "@nestjs/graphql": "^8.0.2",
    "@nestjs/passport": "^8.0.1",
    "@nestjs/platform-express": "^8.0.0",
    "@types/jsonwebtoken": "^8.5.5",
    "@types/mssql": "^7.1.3",
    "apollo-server-express": "^2.25.2",
    "axios": "^0.21.1",
    "crypto": "^1.0.1",
    "graphql": "^15.5.1",
    "joi": "^17.4.2",
    "jsonwebtoken": "^8.5.1",
    "jwa": "^2.0.0",
    "jwks-rsa": "^2.0.4",
    "jwt-decode": "^3.1.2",
    "mssql": "^7.2.0",
    "oauth-1.0a": "^2.2.6",
    "passport": "^0.4.1",
    "passport-custom": "^1.1.1",
    "reflect-metadata": "^0.1.13",
    "rimraf": "^3.0.2",
    "rxjs": "^7.2.0",
    "tedious": "^11.4.0"
  },
  "devDependencies": {
    "@nestjs/cli": "^8.0.0",
    "@nestjs/schematics": "^8.0.0",
    "@nestjs/testing": "^8.0.0",
    "@types/express": "^4.17.13",
    "@types/jest": "^26.0.24",
    "@types/node": "^16.0.0",
    "@types/passport-azure-ad": "^4.3.1",
    "@types/qs": "^6.9.7",
    "@types/supertest": "^2.0.11",
    "@typescript-eslint/eslint-plugin": "^4.28.2",
    "@typescript-eslint/parser": "^4.28.2",
    "eslint": "^7.30.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^3.4.0",
    "jest": "27.0.6",
    "prettier": "^2.3.2",
    "supertest": "^6.1.3",
    "ts-jest": "^27.0.3",
    "ts-loader": "^9.2.3",
    "ts-node": "^10.0.0",
    "tsconfig-paths": "^3.10.1",
    "typescript": "^4.3.5"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".*\\.spec\\.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.(t|j)s"
    ],
    "coverageDirectory": "../coverage",
    "testEnvironment": "node"
  }
}
```

**Comment by mdrideout** [comment]
### Update / Fix
OK so it appears that the minimum instance example platform I provided above was not installing the "old" version of NestJS that I was running which is why that instance was not showing an error.

I've done a bunch of package updates and the problem has been resolved, so it appears to have been fixed at some point already. Here is my new package.json file for reference.

```yaml
{
  "name": "graphql-server",
  "version": "0.0.1",
  "description": "",
  "author": "",
  "private": true,
  "license": "UNLICENSED",
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "nest build",
    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
    "start": "nest start",
    "start:local": "STAGE=local nest start --watch",
    "start:debug": "STAGE=local nest start --debug --watch",
    "start:staging": "STAGE=staging nest start --debug --watch",
    "start:prod": "STAGE=prod node dist/main",
    "start:azure": "SERVER=azure npm run build && func start --port 7072",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "test": "STAGE=local jest",
    "test:watch": "STAGE=local jest --watch --silent",
    "test:watch:verbose": "STAGE=local jest --watch",
    "test:cov": "jest --coverage",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test:e2e": "jest --config ./test/jest-e2e.json"
  },
  "dependencies": {
    "@azure/cosmos": "^3.15.1",
    "@azure/functions": "^1.2.3",
    "@azure/identity": "^1.5.1",
    "@azure/storage-blob": "^12.7.0",
    "@nestjs/apollo": "^10.0.8",
    "@nestjs/axios": "^0.0.7",
    "@nestjs/azure-func-http": "^0.8.0",
    "@nestjs/common": "^8.4.4",
    "@nestjs/config": "^2.0.0",
    "@nestjs/core": "^8.4.4",
    "@nestjs/graphql": "^10.0.8",
    "@nestjs/passport": "^8.2.1",
    "@nestjs/platform-express": "^8.4.4",
    "@types/jsonwebtoken": "^8.5.5",
    "@types/mssql": "^7.1.3",
    "apollo-server-core": "^3.6.7",
    "apollo-server-express": "^3.6.7",
    "axios": "^0.26.1",
    "crypto": "^1.0.1",
    "graphql": "^15.8.0",
    "joi": "^17.4.2",
    "jsonwebtoken": "^8.5.1",
    "jwa": "^2.0.0",
    "jwks-rsa": "^2.0.4",
    "jwt-decode": "^3.1.2",
    "mssql": "^7.2.0",
    "oauth-1.0a": "^2.2.6",
    "passport": "^0.5.2",
    "passport-custom": "^1.1.1",
    "reflect-metadata": "^0.1.13",
    "rimraf": "^3.0.2",
    "rxjs": "^7.5.5",
    "tedious": "^11.4.0"
  },
  "devDependencies": {
    "@nestjs/cli": "^8.2.5",
    "@nestjs/schematics": "^8.0.10",
    "@nestjs/testing": "^8.4.4",
    "@types/express": "^4.17.13",
    "@types/jest": "^26.0.24",
    "@types/node": "^16.0.0",
    "@types/passport-azure-ad": "^4.3.1",
    "@types/qs": "^6.9.7",
    "@types/supertest": "^2.0.11",
    "@typescript-eslint/eslint-plugin": "^4.28.2",
    "@typescript-eslint/parser": "^4.28.2",
    "eslint": "^7.30.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^3.4.0",
    "jest": "^27.0.0",
    "prettier": "^2.3.2",
    "supertest": "^6.1.3",
    "ts-jest": "^27.0.3",
    "ts-loader": "^9.2.3",
    "ts-node": "^10.0.0",
    "tsconfig-paths": "^3.10.1",
    "typescript": "^4.3.5"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".*\\.spec\\.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.(t|j)s"
    ],
    "coverageDirectory": "../coverage",
    "testEnvironment": "node"
  }
}
```

**Comment by AndrewKiri** [comment]
I can confirm that the issue is easily reproducible. All I did is just copy exactly pieces of code from the [doc about union types](https://docs.nestjs.com/graphql/unions-and-enums), and voila:
<img width="1323" alt="Screenshot 2022-05-06 at 20 23 11" src="https://user-images.githubusercontent.com/13384229/167196294-7f1b7fca-c699-438f-a008-a73de5af761b.png">

**Comment by kamilmysliwiec** [comment]
@AndrewKiri you're likely just using an older version of TypeScript in your codebase.

**Comment by AndrewKiri** [comment]
> @AndrewKiri you're likely just using an older version of TypeScript in your codebase.

Yes, exactly! I've updated TS and noticed I was on older version of `nestjs/graphql` which I have also updated to v10, after all version updates everything started to work. 

Thanks a lot 🙏 

Attaching screenshot for anyone interested and with the same problem.

<img width="1419" alt="Screenshot 2022-05-13 at 23 58 20" src="https://user-images.githubusercontent.com/13384229/168394785-c645f241-b428-4a52-8c54-3ac099c54613.png">
